felix/bpf/maps.go

```diff
@@ -594,3 +600,47 @@ func (b *PinnedMap) CopyDeltaFromOldMap() error {
 	}
 	return nil
 }
+
+func (b *PinnedMap) getOldMapVersion() (int, error) {
+	oldVersion := 0
+	dir, name := filepath.Split(b.Filename)
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return 0, fmt.Errorf("error reading pin path %w", err)
+	}
+	for _, f := range files {
+		if strings.Contains(f.Name(), name) {
+			mapName := f.Name()
+			oldVersion, err = strconv.Atoi(string(mapName[len(mapName)-1]))
+			if err != nil {
+				return 0, fmt.Errorf("invalid version %w", err)
+			}
+			if oldVersion < b.Version {
+				return oldVersion, nil
+			}
+			if oldVersion > b.Version {
+				return 0, fmt.Errorf("downgrade not supported %d %d", oldVersion, b.Version)
+			}
+		}
+	}
+	return oldVersion, nil
+}
+
+func (b *PinnedMap) upgrade() error {
+	if b.UpgradeFn == nil {
+		return nil
+	}
+	oldVersion, err := b.getOldMapVersion()
+	if err != nil {
+		return err
+	}
+	// fresh install
+	if oldVersion == 0 {
+		return nil
+	}
+	return b.UpgradeFn(oldVersion, b.Version, b.context)
```

Comment:
this could be where the multiversion/upgrade.go: Upgrade() guts could be and the upgrade.go could go completely, the mock package would only have the maps